dart/example/monster_my_game.sample_generated.dart

```diff
@@ -0,0 +1,456 @@
+// automatically generated by the FlatBuffers compiler, do not modify
+// ignore_for_file: unused_import, non_constant_identifier_names
+
+library my_game.sample;
+
+import 'dart:typed_data' show Uint8List;
+import 'package:flat_buffers/flat_buffers.dart' as fb;
+
+
+class Color {
+  final int value;
+  const Color._(this.value);
+
+  factory Color.fromValue(int value) {
+    if (value == null) return null;
+    if (value < minValue || maxValue < value) {
+      throw new RangeError.range(value, minValue, maxValue);
+    }
+    return values[value - minValue];
+  }
+
+  static const int minValue = 0;
+  static const int maxValue = 2;
+  static bool containsValue(int value) => minValue <= value && value <= maxValue;
+
+  static const Color Red = const Color._(0);
+  static const Color Green = const Color._(1);
+  static const Color Blue = const Color._(2);
+  static get values => [Red,Green,Blue,];
+
+  static const fb.Reader<Color> reader = const _ColorReader();
+
+  @override
+  String toString() {
+    return 'Color{value: $value}';
+  }
+}
+
+class _ColorReader extends fb.Reader<Color> {
+  const _ColorReader();
+
+  @override
+  int get size => 1;
+
+  @override
+  Color read(fb.BufferContext bc, int offset) =>
+      new Color.fromValue(const fb.Int8Reader().read(bc, offset));
+}
+
+class EquipmentTypeId {
+  final int value;
+  const EquipmentTypeId._(this.value);
+
+  factory EquipmentTypeId.fromValue(int value) {
+    if (value == null) return null;
+    if (value < minValue || maxValue < value) {
+      throw new RangeError.range(value, minValue, maxValue);
+    }
+    return values[value - minValue];
+  }
+
+  static const int minValue = 0;
+  static const int maxValue = 1;
+  static bool containsValue(int value) => minValue <= value && value <= maxValue;
+
+  static const EquipmentTypeId NONE = const EquipmentTypeId._(0);
+  static const EquipmentTypeId Weapon = const EquipmentTypeId._(1);
+  static get values => [NONE,Weapon,];
+
+  static const fb.Reader<EquipmentTypeId> reader = const _EquipmentTypeIdReader();
+
+  @override
+  String toString() {
+    return 'EquipmentTypeId{value: $value}';
+  }
+}
+
+class _EquipmentTypeIdReader extends fb.Reader<EquipmentTypeId> {
+  const _EquipmentTypeIdReader();
+
+  @override
+  int get size => 1;
+
+  @override
+  EquipmentTypeId read(fb.BufferContext bc, int offset) =>
+      new EquipmentTypeId.fromValue(const fb.Uint8Reader().read(bc, offset));
+}
+
+/// Vec3
+abstract class Vec3 {
+  double get x;
+  double get y;
+  double get z;
+
+  factory Vec3(List<int> bytes) {
+    fb.BufferContext rootRef = new fb.BufferContext.fromBytes(bytes);
+    return reader.read(rootRef, 0);
+  }
+
+  static const fb.Reader<Vec3> reader = const _Vec3Reader();
+
+}
+
+class _Vec3Impl implements Vec3 {
+  final fb.BufferContext _bc;
+  final int _bcOffset;
+
+  _Vec3Impl(this._bc, this._bcOffset);
+
+  double _x;
+  @override
+  double get x {
```

Comment:
Ok, I'm confused about a bunch of things here:

1. Why does there need to be an abstract class and impl, rather than just 1 class?
2. Why `double`? Are these 32bit in Dart?
3. Does the declaration of `double _x` mean an actual variable is stored? If this an accessor, it should really store zero things, as the data is in the buffer.
4. For my education, what is this `??=` doing?

Reply:
1. I did this mainly because it followed the design in the SDK code.  I'm not necessarily committed to this design, but I suppose it could allow for other implementations of the same model class that aren't FlatBuffer driven more easily (e.g. a `MonsterJsonImpl` or some such thing).
2. `double` is the only floating point type available in Dart and is defined by the spec to be 64 bits (int, per the spec, can be as many bits as you have memory).  There is some notion of lists/vectors of Float32 or Float64s, but not individual numbers that I'm aware of.
3. The declaration `double _x;` is equivalent to `double _x = null;`.  Null is a constant in dart, and as far as I understand does not involve any allocation.  However, I'm pretty sure that internally there'd be a pointer generated for the reference to null.  
4. `??=` is an operator that assigns the variable if it is null, or leaves it as is if it is not null.

Basically, for 3 and 4 - it's caching access to the underlying buffer so that subsequent calls to the getter will not have to touch the underlying buffer again.  I don't have any tests around this, but I'd imagine there are cases where it's faster one way or the other.  This is how it was developed in the SDK code and I kept it.

Reply:
1. in what SDK code? I don't think it makes much sense to inherit from automatically generated classes that represent data in a FlatBuffer.
3. With allocated, I meant stored, sorry. So yes, these variables shouldn't exist.
4. So this is attempting to cache all values? I'd think it should just load from the buffer on each occasion.

Reply:
Ok.  I think this should be plenty fast either way.  I'll get rid of the inherited impl classes too